File: src/test/scala/org/multiverse/benchmark/bank/Account.java
package org.multiverse.benchmark.bank;

import org.multiverse.api.annotations.AtomicMethod;
import org.multiverse.api.annotations.AtomicObject;


/**
 * @author Pascal Felber
 * @since 0.1
 */
@AtomicObject
abstract public class Account {

	static volatile boolean s_disjoint = false;
	static volatile boolean s_yield = false;

	abstract public String getName();

	abstract public float getBalance();

	abstract public void deposit(float amount);

	abstract public void withdraw(float amount) throws OverdraftException;

	@AtomicMethod
	static public void addInterest(Account[] accounts, float rate) {
		for (Account a : accounts) {
			a.deposit(a.getBalance() * rate);
			if (s_yield)
				Thread.yield();
		}
	}

	@AtomicMethod
	static public double computeTotal(Account[] accounts) {
		double total = 0.0;
		for (Account a : accounts) {
			total += a.getBalance();
			if (s_yield)
				Thread.yield();
		}
		return total;
	}

	@AtomicMethod
	static public void transfer(Account src, Account dst, float amount) throws OverdraftException {
		dst.deposit(amount);
		if (s_yield)
			Thread.yield();
		src.withdraw(amount);
	}
}

File: src/test/scala/edu/stanford/ppl/stm/bank/Account.scala
/* CCSTM - (c) 2009-2010 Stanford University - PPL */

// Account

package edu.stanford.ppl.stm.bank


import edu.stanford.ppl.ccstm._

object Account {
  var yieldALot = false

  def addInterest(accounts: Array[Account], rate: Float)(implicit txn: Txn) {
    for (a <- accounts) {
      a.deposit(a.balance.get * rate)
      if (yieldALot) Thread.`yield`
    }
  }

  def computeTotal(accounts: Array[Account])(implicit txn: Txn) = {
    var total = 0.0
    for (a <- accounts) {
      total += a.balance.get
      if (yieldALot) Thread.`yield`
    }
    total
  }

  def transfer(src: Account, dst: Account, amount: Float)(implicit txn: Txn) = {
    dst.deposit(amount)
    if (yieldALot) Thread.`yield`
    src.withdraw(amount)
  }
}

trait Account {
  def name: String
  def balance: Source[Float]
  def deposit(amount: Float)(implicit txn: Txn)
  def withdraw(amount: Float)(implicit txn: Txn)
}
